doc-util/render.libsonnet

```diff
@@ -0,0 +1,324 @@
+        )
+        + root.renderSections(
+          section.subSections,
+          depth + 1,
+          prefixes + [section.name]
+        )
+        for section in sections
+      ])
+    else '',
+
+  renderPackage(package)::
+    (root.templates.package % package)
+    + (
+      if std.length(package.subPackages) > 0
+      then
+        '## Subpackages\n\n'
+        + std.join('\n', [
+          '* [%(name)s](%(path)s)' % {
+            name: sub.name,
+            path: package.name + '/' + sub.name + '.md',
+          }
+          for sub in package.subPackages
+        ]) + '\n\n'
+      else ''
+    )
+    + (root.templates.index % root.index(package.sections))
+    + '\n## Fields\n\n'
+    + root.renderValues(package.values)
+    + root.renderSections(package.sections),
+
+  index(sections, depth=0, prefixes=[])::
+    std.join('\n', [
+      std.join('', [
+        ' '
+        for d in std.range(0, (depth * 2) - 1)
+      ])
+      + (root.templates.sectionLink % {
+           abbr: section.type.abbr,
+           linkName: section.linkName,
+           link:
+             std.asciiLower(
+               std.strReplace(
+                 std.strReplace(root.renderSectionTitle(section, prefixes), '.', '')
+                 , ' ', '-'
+               )
+             ),
+         })
+      + (
+        if std.length(section.subSections) > 0
+        then '\n' + root.index(
+          section.subSections,
+          depth + 1,
+          prefixes + [section.name]
+        )
+        else ''
+      )
+      for section in sections
+    ]),
+
+  sections: {
+    base: {
+      subSections: [],
+      values: [],
+    },
+    object(key, doc, obj, depth):: self.base {
+      name: std.strReplace(key, '#', ''),
+
+      local processed = root.process(obj, depth=depth + 1),
+
+      subSections: processed.sections,
+
+      values: processed.values,
+
+      type: { full: 'object', abbr: 'obj' },
+
+      abbr: self.type.abbr,
+
+      doc:
+        if self.type.full in doc
+        then doc[self.type.full]
+        else { help: '' },
+
+      help: self.doc.help,
+
+      linkName: self.name,
+
+      content:
+        if self.help != ''
+        then self.help + '\n'
+        else '',
+    },
+
+    'function'(key, doc):: self.base {
+      name: std.strReplace(key, '#', ''),
+
+      type: { full: 'function', abbr: 'fn' },
+
+      abbr: self.type.abbr,
+
+      doc: doc[self.type.full],
+
+      help: self.doc.help,
+
+      args: std.join(', ', [
+        if arg.default != null
+        then arg.name + '=' + arg.default
+        else arg.name
+        for arg in self.doc.args
+      ]),
+
+      linkName: '%(name)s(%(args)s)' % self,
+
+      content: '```ts\n%(name)s(%(args)s)\n```\n\n%(help)s' % self,
+
+    },
+
+    value(key, doc, obj):: self.base {
+      name: std.strReplace(key, '#', ''),
+      type: doc.value.type,
+      help: doc.value.help,
+      value: obj,
+    },
+  },
+
+  process(obj, filename='', depth=0)::
```

Comment:
can we have a more descriptive name for this function? It takes some jsonnet and groups it into types of docsonnet objects. But that's not clear in the name `process`.

Reply:
Sure, any suggestions?

Reply:
hah, was hoping you'd work that out :-)

`prepare` would say a bit more than `process`. `prepare_doc_objects` maybe?

Reply:
Renamed it to `prepare`, as it is an internal function it doesn't matter that much.